language/move-prover/bytecode-to-boogie/src/lifetime_analysis.rs

```diff
@@ -0,0 +1,361 @@
+pub enum EdgeType {
+    Local,
+    Global,
+    Field,
+}
+
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub enum Node {
+    Root,
+    Local(LocalIndex),
+}
+
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Edge {
+    edge_type: EdgeType,
+    to: Node,
+}
+
+#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct BorrowGraph {
+    graph: BTreeMap<Node, BTreeSet<Node>>,
+    reverse_graph: BTreeMap<Node, BTreeSet<Node>>,
+    moved_locals: BTreeSet<LocalIndex>,
+}
+
+impl BorrowGraph {
+    pub fn new() -> Self {
+        BorrowGraph {
+            graph: BTreeMap::new(),
+            reverse_graph: BTreeMap::new(),
+            moved_locals: BTreeSet::new(),
+        }
+    }
+
+    // Add an edge in the graph
+    pub fn add_edge(&mut self, from: Node, to: Node) {
+        let mut new_edge_set = BTreeSet::new();
+        new_edge_set.insert(to);
+
+        // Fancy statement that modifies the value of key from if it exists, and inserts otherwise.
+        self.graph
+            .entry(from)
+            .and_modify(|e| {
+                e.insert(to);
+            })
+            .or_insert(new_edge_set);
+
+        // Add edge to reverse map
+        let mut reverse_edge_set = BTreeSet::new();
+        reverse_edge_set.insert(from);
+        self.reverse_graph
+            .entry(to)
+            .and_modify(|e| {
+                e.insert(from);
+            })
+            .or_insert(reverse_edge_set);
+    }
+
+    // Mark a local as moved. Notice that moving the local doesn't mean
+    // it's dead and can be removed
+    pub fn move_local(&mut self, local: LocalIndex) {
+        self.moved_locals.insert(local);
+    }
+
+    // Replace all appearances of before with after in the graph
+    pub fn replace_local(&mut self, before: Node, after: Node) {
+        if self.graph.contains_key(&before) {
+            let to_neighbors = self.graph[&before].clone();
+            for to in &to_neighbors {
+                self.reverse_graph.entry(*to).and_modify(|e| {
+                    e.remove(&before);
+                    e.insert(after);
+                });
+            }
+            self.graph.remove(&before);
+            self.graph.insert(after, to_neighbors);
+        }
+        if self.reverse_graph.contains_key(&before) {
+            let from_neighbors = self.reverse_graph[&before].clone();
+            for from in &from_neighbors {
+                self.graph.entry(*from).and_modify(|e| {
+                    e.remove(&before);
+                    e.insert(after);
+                });
+            }
+            self.reverse_graph.remove(&before);
+            self.reverse_graph.insert(after, from_neighbors);
+        }
+    }
+
+    // Join two borrow graphs, so that self is mutated into a graph consisting of
+    // edges in both graphs
+    pub fn join(&mut self, other: &Self) -> Self {
+        let mut joined = self.clone();
```

Comment:
Doesn't look like you are actually mutating here, but creating a copy. Copying the entire graph for a join might be a bit expensive. In any case, the documentation is misleading, and `&mut self` should not be needed.